### PR TITLE
Add completion for Language Extensions

### PR DESCRIPTION
**Pull request by cronokirby**
First try at resolving #867.

This adds completion for language extensions.

This changes the behaviour of completion generation to only generate completions for language extensions on lines beginning with "{-# LANGUAGE", perhaps we should modify this so it works even if the line just contains "LANGUAGE": `{-#LANGUAGE` and `{-#   LANGUAGE` would then also work.

This adds a `cachedExtensions` field to the `CachedCompletions` record, which is filled via the Ghc api. Going through GhcMod would be much more complicated, since [in ghcmod](https://github.com/DanielG/ghc-mod/blob/master/GhcMod/Exe/Lang.hs#L10) a nice list of language extensions is squashed into a string, which we'd have to parse back. Instead I just get the list via [this](https://hackage.haskell.org/package/ghc-8.6.1/docs/DynFlags.html#v:supportedLanguagesAndExtensions)

I wasn't sure what kind of [completion](https://github.com/alanz/haskell-lsp/blob/2bdec6b34f7fa76b71c4d0931557f64ae9077c75/haskell-lsp-types/src/Language/Haskell/LSP/Types/Completion.hs#L22) to give Language Extensions, so I went with `CiModule` since it seemed the most appropriate

### CODE REVIEW

**Comment by lorenzo** [comment]
I think this is looking good. Could you also add a test? This is the test for the import module completions:

https://github.com/haskell/haskell-ide-engine/blob/master/test/functional/CompletionSpec.hs#L34-L47

It uses this file for comparisons:

https://github.com/haskell/haskell-ide-engine/blob/master/test/testdata/completion/Completion.hs


You can run the functional tests with:

```bash
stack test haskell-ide-engine:func-test
```

**Comment by cronokirby** [comment]
I've added a test to the suite, I've also played around with it on some of my projects and it works quite well

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 1]
```diff
@@ -139,6 +139,12 @@ mkModCompl label =
     Nothing Nothing Nothing Nothing hoogleQuery
   where hoogleQuery = Just $ toJSON $ "module:" <> label
 
+mkExtCompl :: T.Text -> J.CompletionItem
+mkExtCompl label =
+  J.CompletionItem label (Just J.CiModule) Nothing
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: I think changing it to `CiKeyword` makes more sense, but if you think this is just bike shedding, then feel free to disregard :)

```suggestion
  J.CompletionItem label (Just J. CiKeyword) Nothing
```

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs [hunk 2]
```diff
@@ -322,16 +333,21 @@ getCompletions uri prefixInfo (WithSnippets withSnippets) = pluginGetFile "getCo
...
+                             | label <- Fuzzy.simpleFilter fullPrefix list
+                             , enteredQual `T.isPrefixOf` label
+                             ]
+
+      filtImportCompls = filtListWith mkImportCompl importableModules
+      filtExtensionCompls = filtListWith mkExtCompl cachedExtensions
+
+      result
+        | "import " `T.isPrefixOf` fullLine      = filtImportCompls
+        | "{-# LANGUAGE" `T.isPrefixOf` fullLine = filtExtensionCompls
```
Comment (alanz) [inline src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: This is fine for most use cases, but technically the keyword `LANGUAGE` is parsed to be case insensitive.

File: src/Haskell/Ide/Engine/Plugin/HieExtras.hs (hunk 1 above)
Comment (cronokirby) [reply src/Haskell/Ide/Engine/Plugin/HieExtras.hs]: I agree, this does look better.

**Comment by alanz** [comment]
I just tried this, and it works great.

A possible tweak would be to recognise `{-#` and offer to complete with `LANGUAGE` or `OPTIONS_GHC`. and whatever else is valid.

**Comment by lorenzo** [comment]
I’ll create a ticket for that

**Pull request merged**